arborio/swcio.cpp

```diff
@@ -241,272 +206,121 @@ arb::morphology load_swc_arbor(const swc_data& data) {
 }
 
 arb::morphology load_swc_neuron(const swc_data& data) {
-    const auto& records = data.records();
-
-    // Assert that the file contains at least one sample.
-    if (records.empty()) return {};
-
-    const int soma_tag = 1;
-    auto soma_prox = records.front();
-
-    // Assert that root sample has tag 1.
-    if (soma_prox.tag != soma_tag) throw swc_no_soma{soma_prox.id};
-
-    // check for single soma cell
-    bool has_children = false;
+    constexpr int soma_tag = 1;
 
-    // Map of SWC record id to index in `records`.
-    std::unordered_map<int, std::size_t> record_index;
-    record_index[soma_prox.id] = 0;
+    const auto n_samples = data.records().size();
 
-    // Vector of records that make up the soma
-    std::vector<swc_record> soma_records = {soma_prox};
-    int prev_tag = soma_prox.tag;
-    int prev_id = soma_prox.id;
-
-    // Preliminary error checking and building the record_index
-    for (std::size_t i = 1; i < records.size(); ++i) {
-        const auto& r = records[i];
-        record_index[r.id] = i;
+    if (n_samples==0) {
+        return {};
+    }
 
-        if (r.tag == soma_tag) {
-            if (prev_tag != soma_tag)   throw swc_non_consecutive_soma{r.id};
-            if (prev_id != r.parent_id) throw swc_non_serial_soma{r.id};
-            soma_records.push_back(r);
+    // The NEURON interpretation is only applied when the cell has a soma.
+    if (data.records()[0].tag != soma_tag) {
+        const auto& R = data.records();
+        // Search for other soma samples
+        if (auto it=std::find_if(R.begin(), R.end(), [](auto& r) {return r.tag==soma_tag;}); it!=R.end()) {
+            // The presence of a soma tag when there is a non-soma tag at the root
+            // violates the requirement that the parent of a soma sample is also a
+            // soma sample.
+            throw swc_mismatched_tags(it->id);
         }
 
-        // Find record index of the parent
-        auto parent_iter = record_index.find(r.parent_id);
-
-        if (parent_iter == record_index.end() || records[parent_iter->second].id == r.id) throw swc_no_such_parent{r.id};
+        // No soma: fall back to Arbor interpretation.
+        return load_swc_arbor(data);
```

Comment:
This would mean that tags other than 1, 2, 3 and 4 would be allowed in this case. It's not a huge problem but it would be against the specs, right?

Reply:
This was possibly at my insistence. As a general rule, I think it is better not to impose constraints that are not necessary for a consistent and unambiguous interpretation, unless we're trying to validate against the conventions — interpretation and validation are different activities, so to speak. If we writing though, it would be a different story.

Reply:
Okay, but then it's still different behavior depending on whether or not a soma is present in the description.

Reply:
Indeed. I have added a tag check before the call to `load_swc_arbor`.

Reply:
@halfflat, for the NEURON interpretation we explicitly only allow tags = {1,2,3,4}, because of the special way that NEURON interprets these. These get bundled into soma, axon, dendrite and apical "section lists" by NEURON.

Reply:
But, doesn't the reader (our reader) return just a morphology where these 'section lists' correspond to tagged regions?